src/main/java/org/jenkinsci/plugins/workflow/cps/persistence/IteratorHack.java

```diff
@@ -24,91 +24,158 @@
-                    if (fieldName.equals("this$0")) {
-                        try {
-                            Field f = AbstractList.class.getDeclaredField("modCount"); // TODO if not handling only ArrayList.Itr, search in superclasses
-                            f.setAccessible(true);
-                            int modCount = f.getInt(value);
-                            LOGGER.log(Level.FINER, "found a modCount={0}", modCount);
-                            rp.writeField(o, "expectedModCount", modCount, type); // TODO search in superclasses
-                        } catch (Exception x) {
-                            return Futures.immediateFailedFuture(x);
-                        }
-                    }
+        @Override public boolean hasPrevious() {
+            return cursor != 0;
+        }
+        @Override public E previous() {
+            try {
+                if (cursor == 0) {
+                    throw new NoSuchElementException();
                 }
+                cursor--;
+                lastRet = cursor;
+                return list.get(lastRet);
+            } catch (IndexOutOfBoundsException e) {
+                throw new NoSuchElementException();
             }
-            LOGGER.log(Level.FINE, "reconstructed {0}", o);
-            return Futures.immediateFuture(o);
         }
+        @Override public int nextIndex() {
+            return cursor;
+        }
+        @Override public int previousIndex() {
+            return cursor - 1;
+        }
+        @Override public void set(E e) {
+            list.set(lastRet, e);
+        }
+        @Override public void add(E e) {
+            list.add(cursor, e);
+            cursor++;
+            lastRet = -1;
+        }
+    }
+
+    public static <E> ListIterator<E> listIterator(List<E> list) {
+        if (!Caller.isAsynchronous(list, "listIterator") && !Caller.isAsynchronous(IteratorHack.class, "listIterator", list)) {
+            return list.listIterator();
+        }
+        return new ListItr<>(list, 0);
+    }
+
+    public static <E> ListIterator<E> listIterator(List<E> list, int idx) {
+        if (!Caller.isAsynchronous(list, "listIterator", idx) && !Caller.isAsynchronous(IteratorHack.class, "listIterator", list, idx)) {
+            return list.listIterator(idx);
+        }
+        return new ListItr<>(list, idx);
+    }
+
+    /** Serializable replacement for {@link Map#entrySet}. */
+    public static <K, V> Set<Map.Entry<K, V>> entrySet(Map<K, V> map) {
+        if (!Caller.isAsynchronous(map, "entrySet") && !Caller.isAsynchronous(IteratorHack.class, "entrySet", map)) {
+            // In @NonCPS so no need to bother processing this.
+            return map.entrySet();
+        }
+        Set<Map.Entry<K, V>> entries = new LinkedHashSet<>();
+        for (Map.Entry<K, V> entry : map.entrySet()) {
+            // TODO return an object holding references to the map and the key and delegating all calls accordingly
```

Comment:
Yes, see `IteratorHackTest`. Prior to this PR, `ArrayList.iterator` was safe but nothing else (and in particular I had no idea how to support iterating `Map`s and the like—the trick did not generalize well). Now most common interface methods in this category should work, and on any implementation class.